### ISSUE TITLE
How to get forward error in optimizer.py

### ISSUE THREAD

**Issue by lincoln2010** [apache/incubator-mxnet#3546]
Is there are anyone know how to get forward error in optimizer.py, as currently we could only get gradient results in optimizer.py. Many thanks in advance.

**Comment by lincoln2010** [comment]
PS: To be more specifically, how to get forward error like in the function of update:
 def update(self, index, weight, grad, state):
        """Update the parameters.

        Parameters
        ----------
        index : int
            An unique integer key used to index the parameters

        weight : NDArray
            weight ndarray

        grad : NDArray
            grad ndarray

        state : NDArray or other objects returned by init_state
            The auxiliary state used in optimization.
        """

**Comment by piiswrong** [comment]
error is computed by metric, not optimizer

**Comment by lincoln2010** [comment]
It seems that error is computed by the loss function not the metric. And the problem is that how can we get the error in optimizer.py in each update.

**Comment by lincoln2010** [comment]
PS: At the end of model definition, usually there would be loss layers like softmax but we can define a different metrics to display the errors. In the optimization process, the error value from the loss layers are used to compute the gradient. Is that right?

**Comment by piiswrong** [comment]
gradient is computed by layers. Optimizer only update it. If you want to do something special with the gradient it should be done in a custom loss layer

**Comment by lincoln2010** [comment]
Thank you very much for your reply. Do you have any suggestions if we would like to get the error value in the optimizer.py  from a custom loss layer? @piiswrong

**Comment by piiswrong** [comment]
keep a reference to your model in optimizers

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
